src/txadmission.cpp

```diff
@@ -270,51 +270,45 @@ void LimitMempoolSize(CTxMemPool &pool, size_t limit, unsigned long age)
 
 void CommitTxToMempool()
 {
+    // Committing the tx to the mempool takes time.  We can continue to validate non-conflicting tx during this time.
+    // To do so, before the transactions are finally commited to the mempool the txCommitQ pointer is copied
+    // to txCommitQFinal so that the lock on txCommitQ can be released and processing can continue.
+    // However, the incomingConflicts detector is not reset until all the transactions are committed to the mempool.
+    std::map<uint256, CTxCommitData> *txCommitQFinal = nullptr;
+
     std::vector<uint256> vWhatChanged;
     {
-        boost::unique_lock<boost::mutex> lock(csCommitQ);
-        avgCommitBatchSize = (avgCommitBatchSize * 24 + txCommitQ->size()) / 25;
-        LOG(MEMPOOL, "txadmission committing %d tx\n", txCommitQ->size());
+        // We must hold the mempool lock for the duration because we want to be sure that we don't end up
+        // doing this loop in the middle of a reorg where we might be clearing the mempool.
+        WRITELOCK(mempool.cs_txmempool);
```

Comment:
Acutally now that I remember, this is why we made commitqfinal a global.  Because during a  re-org we HAVE to clear the commitqfinal. If we don't clear it, and equeue the transactions again, we risk getting decendants which have no parents pushed into the mempool and when we try to build a block from them we will crash.

Reply:
Ok, never mind.  The locking here is actually ok. We don't have to clear the commitqfinal during a re-org because the mempool here is locked throughout.